### PR TITLE
chore(assertions): release as an alpha module in CDKv2

### LINKED ISSUE #15589: Handle the `@aws-cdk/assertions` library in the V2 build of unstable modules
The `@aws-cdk/assertions` library uses a [build process](https://github.com/aws/aws-cdk/blob/dd38eff318c31bf2c5308f6b3daae0fd433b4370/packages/%40aws-cdk/assertions/package.json#L54) to "vendor in" a few dependencies (`@aws-cdk/cfnspec` and `@aws-cdk/assert`, I believe). This makes it incompatible with the V2 build of experimental modules added in https://github.com/aws/aws-cdk/pull/15442. We need to fix that process.

This will also unblock including its dependents, like `@aws-cdk/aws-synthetics` (excluded [here](https://github.com/aws/aws-cdk/blob/a2e95e97fc775d061b405fdb932ffda269084357/packages/%40aws-cdk/aws-synthetics/package.json#L112-L112)).

### PR DESCRIPTION
**Pull request by skinny85**
The `@aws-cdk/assertions` library was previously excluded from the V2 build of the experimental modules,
because the `vendor.sh` script it used for vendoring in a few modules
(`cfnspec`, `cloudformation-diff`, and `assert-internal`)
used paths relative to the V1 module inside the repository.

Change the `vendor.sh` script to short-circuit and exit when running in `transform` phase for
the alpha packages. That's required because the transform phase re-writes imports when copying
the TS files (because it needs to know which modules are also experimental to do the rewriting correctly),
but the `vendor.sh` script runs only after the `build` script for the package has been invoked,
which means any TS files copied by it successfully would _not_ have their imports re-written.
Since the build for alpha packages runs after the build for `@aws-cdk/assertions`
(we add an explicit dependency to make sure that happens),
those files are already "vendored in" correctly, and will be copied when creating the separate package.

Also adjust the module copying logic to account for bundled dependencies,
which `@aws-cdk/assertion` uses.

Closes #15589

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (nija-at) [inline packages/@aws-cdk/assertions/vendor-in.sh:28]: Can you copy or re-write this comment as written in the PR description? The comment that's in the code is not clear.

**Review by nija-at — CHANGES_REQUESTED** [review]
All of this 'feels' super hacky. It feels too easy to introduce defects, unintended side effects, and worse of all, debugging is going to be a nightmare.

How about a different approach?

Move 'cloudformation-diff' into its own repo outside of the monorepo. It depends on a small part of 'cfnspec'. Without diving too deep into it, I think it should be possible to refactor this part of cfnspec into cloudformation-diff as well.

Then, copy over the files from 'assert-internal' into 'assertions' as part of the source tree instead of copying them in.

All that will remain to be copied over would the cloudformation spec from cfnspec which can then be passed as input to the 'cloudformation-diff' module.

cc @eladb @rix0rrr

**Comment by skinny85** [comment]
@nija-at before I reply, let me understand something.

Your proposal is to actually change the way `@aws-cdk/assertions` is built completely, right? This is not only a change in V2, it would also apply to the V1 version of `@aws-cdk/assertions` - am I understanding you correctly?

**Comment by nija-at** [comment]
> This is not only a change in V2, it would also apply to the V1 version of `@aws-cdk/assertions` - am I understanding you correctly?

Yes. Remove further v2 special casing.

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (skinny85) [reply packages/@aws-cdk/assertions/vendor-in.sh]: Done.

**Comment by skinny85** [comment]
> > This is not only a change in V2, it would also apply to the V1 version of `@aws-cdk/assertions` - am I understanding you correctly?
> 
> Yes. Remove further v2 special casing.

OK. Can we get this PR merged then? I'm fine if you want to change how `@aws-cdk/assertions` is built, but I'd rather not block V2 experimental modules on it.

**Comment by nija-at** [comment]
I'm not comfortable approving this change. We should not be looking at these independently.

This change is increasing the complexity of our build system, which is already overloaded and complicated.
All solutions must strive to simplify, otherwise we're headed towards it being unmanageable.

Given that this is a v2 branch only change, this moves the burden of debugging and resolution to the oncall or somewhere later from the PR that introduces any defect here.
For the reasons stated in my previous comment, the debugging of this is going to be hard, with all the special cases peppered around.

We need to find a solution that makes the overall system simpler.

**Comment by nija-at** [comment]
Two options that come to mind are - the approach I had mentioned above, or supporting dependency bundling within the monorepo so we don't have to vendor in.

Maybe there are others?

**Comment by skinny85** [comment]
Not sure why we can't enable this for V2 experimental builds while working on improving the mess in `@aws-cdk/assertions` concurrently? Whatever change you want to do to the build of `@aws-cdk/assertions`, it will be a completely separate PR from this one, targeting the `master` branch. If you want to get rid of `vendor.sh` script, this means the workaround that I added here will be removed anyway, so I don't think introducing it temporarily is a big deal. This entire change in `@aws-cdk/assertions` is 4 lines of Bash.

I don't think making this PR wait for the PR with the build changes for `@aws-cdk/assertions` is a good idea. We are coupling things together that don't have to be coupled, and so we are slowing down the overall development speed of the project. In fact, you're basically saying that this PR will be blocked until a different PR, which has not yet even been opened, is merged. That means you're introducing a new dependency to the critical path of releasing the experimental modules project. That can't be a good thing.

> Given that this is a v2 branch only change, this moves the burden of debugging and resolution to the oncall or somewhere later from the PR that introduces any defect here.
> For the reasons stated in my previous comment, the debugging of this is going to be hard, with all the special cases peppered around.

This is already the case, whether we merge this PR or not. I take the responsibility for fixing these issues, not the oncall, like I already did in https://github.com/aws/aws-cdk/pull/15648.

**Comment by rix0rrr** [comment]
> Since the build for @aws-cdk-lib runs after the build for @aws-cdk/assertions
(we add an explicit dependency to make sure that happens),
those files are already "vendored in" correctly,
and will be copied when creating the separate package.

This seems to suggest that it's fine to NOT run `vendor.sh` because everything will be okay already.

If so, then why is the fix not to do the following in the monorepo preparation script?

```
ln -sf /bin/ture vendor.sh 
```

?

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (rix0rrr) [inline packages/@aws-cdk/assertions/vendor-in.sh:25]: Here's a 💯 nitpick, but this is not what "fail closed" means. "Fail closed" means to NOT proceed with something if you're unsure about the current state of the system. I think this is more akin to "fail open", proceed regardless.

And it's something we probably shouldn't be doing either way :)

**Comment by skinny85** [comment]
> > Since the build for @aws-cdk-lib runs after the build for @aws-cdk/assertions
> > (we add an explicit dependency to make sure that happens),
> > those files are already "vendored in" correctly,
> > and will be copied when creating the separate package.
> 
> This seems to suggest that it's fine to NOT run `vendor.sh` because everything will be okay already.
> 
> If so, then why is the fix not to do the following in the monorepo preparation script?
> 
> ```
> ln -sf /bin/true vendor.sh 
> ```
> 
> ?

It's a possible solution, but I'd rather keep these hacks locally to the packages, instead of putting them in the experimental packages generation. For example, if Niranjan's plan goes through, and the build of `@aws-cdk/assertions` is changed to no longer use the `vendor.sh` script, this hack will be naturally removed, while it would stay in indefinitely if it was part of the experimental packages build.

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (skinny85) [reply packages/@aws-cdk/assertions/vendor-in.sh]: Good callout. I changed it to just short-circuit instead.

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (nija-at) [inline packages/@aws-cdk/assertions/vendor-in.sh:21]: can we base this on an environment variable such as `SKIP_GEN` or such that will be enabled during the new phase we plan to introduce?

Comment (nija-at) [reply packages/@aws-cdk/assertions/vendor-in.sh:21]: Thought about this a bit more. Given that we decided to add a new phase that runs after the 'build' phase, the env var should set `XXX_PHASE=true` or something like that. We should decide what to call this phase.

**Comment by skinny85** [comment]
@nija-at I have rebased this PR on top of https://github.com/aws/aws-cdk/pull/15722. After https://github.com/aws/aws-cdk/pull/15722 has been merged, if you agree with the changes here, if you could change the base of the PR to be `v2-main`, and then approve it, it _should_ apply cleanly. Thanks!

File: packages/@aws-cdk/assertions/vendor-in.sh
Comment (skinny85) [reply packages/@aws-cdk/assertions/vendor-in.sh]: Done.

File: tools/pkglint/lib/rules.ts
Comment (nija-at) [inline tools/pkglint/lib/rules.ts:1378]: Why is this being skipped?

File: tools/individual-pkg-gen/copy-files-removing-deps.ts
Comment (nija-at) [inline tools/individual-pkg-gen/copy-files-removing-deps.ts:180]: Should we change this directly on the v2 branch here - https://github.com/aws/aws-cdk/blob/master/tools/cdk-build-tools/config/eslintrc.js, and drop all this extra logic ?

File: tools/pkglint/lib/rules.ts
Comment (njlynch) [reply tools/pkglint/lib/rules.ts:1378]: Huh, it's not clear. I removed this check and re-ran pkglint, and didn't see any differences in any of these packages. Perhaps a hold-out from an older iteration? I'll remove for now until the reason becomes clearer.

File: tools/individual-pkg-gen/copy-files-removing-deps.ts
Comment (njlynch) [reply tools/individual-pkg-gen/copy-files-removing-deps.ts:180]: That would change the rules for *all* packages in v2, not just the alpha packages, correct? These alpha packages have slightly different dependency setups, which requires the customization to the rule.

File: tools/pkglint/lib/rules.ts
Comment (njlynch) [reply tools/pkglint/lib/rules.ts]: Huh, I must not have built and/or linted the right thing. Here's the build failure.

```
@aws-cdk-lib-alpha/assertions: In package package.json
@aws-cdk-lib-alpha/assertions: - [yarn/nohoist-bundled-dependencies] Repository-level 'workspaces.nohoist' directive is missing: @aws-cdk-lib-alpha/assertions/colors, @aws-cdk-lib-alpha/assertions/colors/**, @aws-cdk-lib-alpha/assertions/diff, @aws-cdk-lib-alpha/assertions/diff/**, @aws-cdk-lib-alpha/assertions/fast-deep-equal, @aws-cdk-lib-alpha/assertions/fast-deep-equal/**, @aws-cdk-lib-alpha/assertions/string-width, @aws-cdk-lib-alpha/assertions/string-width/**, @aws-cdk-lib-alpha/assertions/table, @aws-cdk-lib-alpha/assertions/table/** (fixable)
```

Comment (nija-at) [reply tools/pkglint/lib/rules.ts]: I don't know enough to understand this. What is this rule meant to prevent and why we should not apply this to v2?

For example, an alternate solution could be to add this to the 'nohoist' flag here - https://github.com/aws/aws-cdk/blob/v2-main/packages/individual-packages/lerna.json. 
I'm not saying that's the right solution, just that I don't understand enough of it.

Can you provide a little more context?

File: tools/individual-pkg-gen/copy-files-removing-deps.ts
Comment (nija-at) [reply tools/individual-pkg-gen/copy-files-removing-deps.ts:71]: Ok, I'm fine with this.

File: tools/pkglint/lib/rules.ts
Comment (njlynch) [reply tools/pkglint/lib/rules.ts]: It's really not clear. I don't think I understand why Adam wanted to exclude the alpha packages from this rule.

The build is complaining that a bundled dependency (e.g., colors) is not added to the top-level `nohoist` config, which keeps those dependencies at the local package level rather than hoisting them up to the top-level workspace. We have colors as a nohoist listing three other times already (`@aws-cdk/assertions/colors`, `aws-cdk-lib/colors`, and `monocdk/colors`). Not sure why we wouldn't add the alpha packages in the same way.

I ran pkglint -f in the alpha assertions package, which added the relevant entries to the top-level package.json. This should solve the build error, and I think is the right thing to do.

Comment (nija-at) [reply tools/pkglint/lib/rules.ts]: Thanks for diving in. Sounds like the right approach, until we have info that says otherwise.

**Review by nija-at — APPROVED** [review]
Thanks for taking care of this @njlynch 

I've updated the title to be cleaner. It looks the description is slightly stale (e.g., uses the term @aws-cdk-lib while we've changed to @aws-cdk-lib-alpha).

Would be good to clean up the description before merging.

**Pull request merged**
